core/templates/services/question-validation.service.ts

```diff
@@ -53,6 +53,24 @@ export class QuestionValidationService {
     );
   }
 
+  areAnswerGroupsNotEqual(question: Question): boolean {
+    const answerGroups = question._stateData.interaction.answerGroups;
+    const ruleSet = new Set<string>();
+
+    for (const group of answerGroups) {
+      for (const rule of group.rules) {
+        const ruleKey = `${rule.type}:${JSON.stringify(rule.inputs)}`;
```

Comment:
Is the rule.inputs dict guaranteed to have a particular order? If not then applying JSON.stringify on the same object in different contexts may not produce the same result.

Also maybe it's better to consider this a ruleHash (it's not so much a "key", it's a string encoding of a rule).